### ISSUE TITLE
Enhance support for adding SubjectAlternativeNames extension

### ISSUE THREAD

**Issue by hanno-arm** [ARMmbed/mbedtls#1017]
__Task:__ Implement functionality to add the `SubjectAlternativeName` extension in a CSR through some human readable input format. Currently, only the generic extension adding function `mbedtls_x509write_csr_set_extension` is available, which needs an ASN.1 formatted input.

__Relates to:__ Forum issue https://tls.mbed.org/discussions/crypto-and-ssl/sample-to-create-a-x509-certificate-with-alternate-name

**Comment by ciarmcom** [comment]
ARM Internal Ref: IOTSSL-1557

**Comment by RonEld** [comment]
Related to #459

**Comment by doru91** [comment]
What is the status for this enhancement?

**Comment by hanno-arm** [comment]
@doru91, we haven't yet found the time to work on this.

**Comment by snej** [comment]
mbedTLS also can't parse the SubjectAlternativeName (or other v3 extensions like ns_cert_type) when reading a CSR from data. This is a bummer if you're trying to use mbedTLS as a CA, i.e. signing other people's CSRs.

I have a patch for 2.19.1 that addresses both these issues. I'll try to post a PR soon.

**Comment by mprse** [comment]
Linking PR that address this issue: https://github.com/Mbed-TLS/mbedtls/pull/6852.

**Comment by mprse** [comment]
I analysed implementation provided in https://github.com/Mbed-TLS/mbedtls/pull/6852 and from my perspective it is partialy incorrect. I need confirmation to clarify what exactly we want to do in this task.

Current implementation adds support for writing `SubjectAlternativeName` to CSR. The following types of `GeneralNames` are supported:
- otherName                              [0]     OtherName
- dNSName                                [2]     IA5String
- uniformResourceIdentifier       [6]     IA5String

Unfortunately current implementation is treating each of above types in the same way. It is correct for `IA5String`, but I think `OtherName` should be handled differently. 

[rfc5280](https://www.rfc-editor.org/rfc/rfc5280):
```
   OtherName ::= SEQUENCE {
        type-id    OBJECT IDENTIFIER,
        value      [0] EXPLICIT ANY DEFINED BY type-id }
```

Now, looking at the current code we are supporting only `HardwareModuleName` type of `OtherName`:

https://github.com/Mbed-TLS/mbedtls/blob/ab1f3c153a90d6943c174613b715d19dfddfc83a/include/mbedtls/x509.h#L254-L301

[New ASN.1 Modules for Cryptographic Message Syntax (CMS) and S/MIME](https://datatracker.ietf.org/doc/html/draft-ietf-smime-new-asn1-07):
```
  on-hardwareModuleName OTHER-NAME ::=
          { HardwareModuleName IDENTIFIED BY id-on-hardwareModuleName }
  id-on-hardwareModuleName OBJECT IDENTIFIER ::= {
      iso(1) identified-organization(3) dod(6) internet(1) security(5)
      mechanisms(5) pkix(7) on(8) 4 }

  HardwareModuleName ::= SEQUENCE {
      hwType OBJECT IDENTIFIER,
      hwSerialNum OCTET STRING }
```

With current implementation generated CSR with `dNSName` and `otherName` looks as follows:
![image](https://user-images.githubusercontent.com/30721012/218326690-72f4c457-890e-4518-907d-9fe7e1486f6e.png)

This is how openssl generated CSR with `dNSName` and `otherName(hardwareModuleName )`
![image](https://user-images.githubusercontent.com/30721012/218326787-e8d00496-8456-4bad-a496-30465592fe6b.png)


1. We need special implementation for `SubjectAlternativeName /OtherName/hardwareModuleName` type? It should generate CSR consistent with the openssl example which seems to be consistent with spec.
2. Only `hardwareModuleName` type of `OtherName` should be supported(at least for now)?
3. How input string should be formatted?
 `otherName:1.3.6.1.5.5.7.8.4;1.3.6.1.4.1.17.3;123456` where:
- `1.3.6.1.5.5.7.8.4`: OID of `hardwareModuleName` (only this value supported)
- `1.3.6.1.4.1.17.3`: OID of `hwtype` (any value)
- `123456`: `hwserial`(octet string, any value)

Looks like we also need function to convert OID string (e.g. `1.3.6.1.4.1.17.3`) to array of bytes. We have function that does the opposite transition (`mbedtls_oid_get_numeric_string`).

**Comment by davidhorstmann-arm** [comment]
> Looks like we also need function to convert OID string (e.g. `1.3.6.1.4.1.17.3`) to array of bytes. We have function that does the opposite transition (`mbedtls_oid_get_numeric_string`).

I started working on such a function a little while ago when working on parsing Distinguished Names. If it helps, I can dig it out and submit it, shouldn't take too long.

**Comment by mprse** [comment]
> I started working on such a function a little while ago when working on parsing Distinguished Names. If it helps, I can dig it out and submit it, shouldn't take too long.

Oh that would be very helpful. Thanks!

**Comment by yanesca** [comment]
As far as I can tell the only `subjectAltName` types we need for OPC UA are `uniformResourceIdentifier`, `dNSName` and `IPAddress`:
https://reference.opcfoundation.org/Core/Part6/v105/docs/6.2

Are we sure we need `OtherName` types? What am I missing?

**Comment by mprse** [comment]
> Are we sure we need OtherName types? What am I missing?

The original [PR](https://github.com/Mbed-TLS/mbedtls/pull/6852) is adding support for `URI`, `DNS` and `OtherName` (`cert_req.c`). `OtherName` part is handled incorrectly as described above.
My plan was to support all names that we are parsing(this includes `OtherName` and `DirectoryName` in near future).

If we decide to relay on provided document and handle only  `URI`, `DNS` and `IP` this will simplify things as all of them are of `IA5String` type.

BTW I have already implementation for `otherName/hardwareModuleName ` ready, so I guess we can use it as additional feature?

**Comment by yanesca** [comment]
> BTW I have already implementation for otherName/hardwareModuleName  ready, so I guess we can use it as additional feature?

Of course, I think the requirements here are:
1. adopt and merge the original PR
2. make sure it has the features OPC-UA needs

So it is justified to have `otherName`, and `hardwareModuleName` sounds like a good option to showcase/test it.

**Comment by mprse** [comment]
Ok I will change the PR to support `uniformResourceIdentifier`, `dNSName` and `IPAddress` and introduce `hardwareModuleName` in the separate PR.

**Comment by davidhorstmann-arm** [comment]
Unsure if this is still needed, but I've submitted the OID parsing function anyway as #7106.

In the course of tidying I found some small bugs in our OID-to-string function `mbedtls_oid_get_numeric_string`. Fixes are in #7105, ideally need to be merged first.

**Comment by mprse** [comment]
> Unsure if this is still needed, but I've submitted the OID parsing function anyway as https://github.com/Mbed-TLS/mbedtls/pull/7106.

Thanks @daverodgman for OID parsing function. I will add support for `otherName` in separate PR (and use your OID parsing function).

**Comment by davidhorstmann-arm** [comment]
> Thanks @daverodgman for OID parsing function. I will add support for `otherName` in separate PR (and use your OID parsing function).

Excellent! Actually I'm @davidhorstmann-arm, but I appreciate it's easy to get mixed up :)

**Comment by mprse** [comment]
> Actually I'm @davidhorstmann-arm, but I appreciate it's easy to get mixed up :)

I'm sorry @davidhorstmann-arm  I confirmed suggested hint too fast.

**Comment by mprse** [comment]
Closing per https://github.com/Mbed-TLS/mbedtls/pull/7009.
